AX: input[type=date] individual fields are announced as "group"

Christop406 opened the pull request:
#### d4f4aedd06563422f36fff424b3592646262e0f3
<pre>
AX: input[type=date] individual fields are announced as &quot;group&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=251544">https://bugs.webkit.org/show_bug.cgi?id=251544</a>
rdar://problem/104928713

Reviewed by Aditya Keerthi.

This fixes the bug by adding labels representing the &quot;day&quot;, &quot;month&quot;, and &quot;year&quot; inputs so that a screen reader has something to read,
instead of just &quot;group&quot; - it also switches the &quot;role&quot; attribute of the same divs to &quot;spinbutton&quot; to more accurately
reflect how they are used.

* LayoutTests/accessibility/date-input-aria-label-and-value-expected.txt: Added.
* LayoutTests/accessibility/date-input-aria-label-and-value.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
(WebCore::DateTimeDayFieldElement::create):
(WebCore::DateTimeMonthFieldElement::create):
(WebCore::DateTimeSymbolicMonthFieldElement::create):
(WebCore::DateTimeYearFieldElement::create):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
(WebCore::DateTimeNumericFieldElement::setEmptyValue):
(WebCore::DateTimeNumericFieldElement::setValueAsInteger):
(WebCore::DateTimeNumericFieldElement::setARIAValueAttributesWithInteger):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXDateFieldMonthText):
(WebCore::AXDateFieldDayText):
(WebCore::AXDateFieldYearText):
* Source/WebCore/platform/LocalizedStrings.h:

Canonical link: <a href="https://commits.webkit.org/261123@main">https://commits.webkit.org/261123@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/62ae5663fb8f8f8c405f63b5fb89bbe259d231bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110578 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19658 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1934 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119446 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114534 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10790 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102874 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116320 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15703 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98884 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43995 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97663 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85909 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12311 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31889 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12873 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8853 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18237 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51505 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7707 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14755 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -1911,4 +1920,3 @@
 
 /* Option in segmented control for choosing list type in text editing */
 "•" = "•";
-
```
Please undo this change.

Christop406 replied:
Not even sure how i did this lol, removed

shivamidow commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
 Ref<DateTimeMonthFieldElement> DateTimeMonthFieldElement::create(Document& document, FieldOwner& fieldOwner)
 {
-    auto element = adoptRef(*new DateTimeMonthFieldElement(document, fieldOwner));
+    auto dateEl = new DateTimeMonthFieldElement(document, fieldOwner);
```
/* Unofficial Review */
dateEl -> dateElement? https://webkit.org/code-style-guidelines/#names-basic

Christop406 replied:
Removed this extra element per other comments. Was a remnant of me trying to debug something.

shivamidow commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
+    auto dateEl = new DateTimeMonthFieldElement(document, fieldOwner);
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
+    auto element = adoptRef(*dateEl);
```
/* Unofficial Review */
Is there a reason why we need dateEl here? Simply using element seems fine.
element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString...);

Christop406 replied:
Nope, removed :)

pxlcoder commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -49,6 +51,8 @@ Ref<DateTimeDayFieldElement> DateTimeDayFieldElement::create(Document& document,
     static MainThreadNeverDestroyed<const AtomString> dayPseudoId("-webkit-datetime-edit-day-field"_s);
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(dayPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Can we get the "spinbutton" string using `accessibilityRoleToString` from `AccessibilityObjectInterface.h`?

twilco replied:
I think `accessibilityRoleToString` may not be what we want, as that function is more for debugging and exposes things that don't actually map to real ARIA roles.

We do have a mapping from WebCore roles to ARIA roles via `AccessibilityObject::reverseAriaRoleMap()`, but that's not exposed outside the class right now. In my opinion hardcoding "spinbutton" is OK here, but if you feel we should do this as a lookup, then the right way to do it would be exposing a public static function (i.e. `AccessibilityObject::webCoreRoleToAriaRole(AccessibilityRole)`) that uses this reverse role map.

shivamidow commented on Source/WebCore/platform/LocalizedStrings.cpp:
```diff
@@ -719,6 +719,21 @@ String AXTimeFieldText()
 
+String AXDateFieldMonthText()
+{
+    return WEB_UI_STRING("month", "accessibility label for a date field month input.");
```
/* Unofficial Review */
I suggest you to match these strings with others like "accessibility role description for XXX."

pxlcoder commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -219,7 +223,10 @@ DateTimeMonthFieldElement::DateTimeMonthFieldElement(Document& document, FieldOw
-    auto element = adoptRef(*new DateTimeMonthFieldElement(document, fieldOwner));
+    auto dateEl = new DateTimeMonthFieldElement(document, fieldOwner);
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    dateEl->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Can you do this after the call to `initialize` below?

And undo the change to the variable name.

Christop406 replied:
yep, both done

pxlcoder commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -278,6 +285,8 @@ Ref<DateTimeSymbolicMonthFieldElement> DateTimeSymbolicMonthFieldElement::create
     static MainThreadNeverDestroyed<const AtomString> monthPseudoId("-webkit-datetime-edit-month-field"_s);
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(monthPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldMonthText() });
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
```
Let's be consistent with the order we set this in. Either label, then role, or vice versa.

Christop406 replied:
updated

pxlcoder commented on Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
+    setValueText(value);
+}
+
+void DateTimeNumericFieldElement::setValueText(int value)
```
I would rename this to `setARIAValueTextWithInteger` to be more explicit.

shivamidow commented on Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
```diff
@@ -59,6 +59,7 @@ class DateTimeNumericFieldElement : public DateTimeFieldElement {
     void setValueAsInteger(int, EventBehavior = DispatchNoEvent) final;
     void stepDown() final;
     void stepUp() final;
+    void setValueText(int);
```
/* Unofficial Review */
This method is confusing. We set a text value but take int? Maybe, we can change it to something like setValueText(String&).

Christop406 replied:
Updated the name to be more representative of what it does, per another comment on this PR.

twilco commented on Source/WebCore/platform/LocalizedStrings.cpp:
```diff
@@ -719,6 +719,21 @@ String AXTimeFieldText()
+
+String AXDateFieldDayText()
+{
+    return WEB_UI_STRING("day", "accessibility label for a date field month input.");
```
Seems like a copy-paste error here —  "month input" should probably be "day input". Same issue for `AXDateFieldYearText()`.

twilco commented on Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
+
+void DateTimeNumericFieldElement::setValueText(int value)
+{
+    setAttributeWithoutSynchronization(HTMLNames::aria_valuetextAttr, AtomString { WTF::String::number(value) });
```
I think you can use `String::number(value)` rather than `AtomString { WTF::String::number(value) }` here.

Christop406 replied:
`String::number(value)` didn't work but `AtomString::number(value)` did

twilco commented on Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
```diff
@@ -117,19 +118,27 @@ void DateTimeNumericFieldElement::setEmptyValue(EventBehavior eventBehavior)
 {
     m_typeAheadBuffer.clear();
     setValueAsInteger(value, DispatchInputAndChangeEvents);
+    setValueText(value);
```
Is this call to `setValueText` necessary? The line above calls `setValueAsInteger`, which calls `setValueText`, which I believe makes the call on this line redundant.

Christop406 replied:
Good catch!

twilco reviewed (commented):
Thanks for working on this!

It would be nice to have a layout test to ensure future changes don't break accessibility. Two cases come to mind:

1. Put keyboard focus onto a date field and send up / down keys, and confirm the accessibility label is what we expect
2. Update the date input's value via JS, and again verify the accessibility label of the input as a whole and the sub-fields within

You can read about layout tests here:

https://github.com/WebKit/WebKit/blob/main/Introduction.md#running-layout-tests

You'll want to add your new test to `LayoutTests/accessibility`.

Within the context of layout tests, you can use a global `eventSender` object to simulate events. So for the first testcase, you may want something like this:

```
// Double childAtIndex to get through `<div pseudo="-webkit-datetime-edit">` to `<div pseudo="-webkit-datetime-edit-fields-wrapper">`
var axDateFieldWrapper = accessibilityController.accessibleElementById("date-input").childAtIndex(0).childAtIndex(0);

document.getElementById("date-input").focus();
// Focus should be on the first of the three groups (e.g. the "month" field)

eventSender.keyDown("upArrow");
// Verify the description. aria-valuetext is exposed through AccessibilityNodeObject::valueDescription(), which is in turn exposed through AccessibilityUIElement::valueDescription() for layout tests.
testOutput += expect("axDateFieldWrapper.childAtIndex(0).valueDescription", "AXValueDescription: ...some expected value...");

// Maybe try some more up / down arrow keys, and verify appropriately

// Tab to the next field and repeat
eventSender.keyDown("\t");
...
```

In terms of general test structure, https://github.com/WebKit/WebKit/blob/1b7d7ef4658c992e734bb24c4a3c04c5e4a01ad7/LayoutTests/accessibility/aria-multiline.html is a good example.

twilco commented on Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
```diff
@@ -49,6 +51,8 @@ Ref<DateTimeDayFieldElement> DateTimeDayFieldElement::create(Document& document,
     ScriptDisallowedScope::EventAllowedScope eventAllowedScope { element };
     element->initialize(dayPseudoId);
+    element->setAttributeWithoutSynchronization(HTMLNames::roleAttr, AtomString { "spinbutton"_s });
+    element->setAttributeWithoutSynchronization(HTMLNames::aria_labelAttr, AtomString { AXDateFieldDayText() });
```
Seems like this spot still sets aria-label after the role unlike the other two methods.

twilco commented on Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
```diff
@@ -59,6 +59,7 @@ class DateTimeNumericFieldElement : public DateTimeFieldElement {
     void setValueAsInteger(int, EventBehavior = DispatchNoEvent) final;
     void stepDown() final;
     void stepUp() final;
+    void setARIAValueTextWithInteger(int);
```
Can this be private?

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,84 @@
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
+    var testOutput = "Tests that aria-label & aria-valuetext update correctly for date inputs.\n\n";
```
Please use "and" instead of & here ([WebKit coding style](https://webkit.org/code-style-guidelines/) prefers to spell things out).

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,84 @@
+        .childAtIndex(0)
+        .childAtIndex(0);
+
+    function expectChildToHaveDescription(childIdx, descText) {
```
Please use "childIndex" and "descriptionText" (or "description"), or some other fully spelled out terms.

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,84 @@
+        .childAtIndex(0);
+
+    function expectChildToHaveDescription(childIdx, descText) {
+        return expect("axDateFieldWrapper.childAtIndex(" + childIdx + ").valueDescription", "'AXValueDescription: " + descText + "'");
```
You can use JavaScript template literals here:

```
expect(`axDateFieldWrapper.childAtIndex(${childIndex}).valueDescription`, `"AXValueDescription: ${description}"`);
```

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,84 @@
+    function keyDown(key, repeat = 1) {
+        for (let i = 0 ; i < repeat ; i++) {
+            eventSender.keyDown(key);
+        }
```
The curly braces aren't necessary here.

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,84 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
```
Nice work on the test! Some high-level comments on style:

1. Could we reduce the amount of newlines throughout this test? I personally like to group things by pairing impetus with expectation, so concretely that might look something like:

```
const dateInput = document.getElementById("date-input");
dateInput.focus();
keyDown("upArrow");
testOutput += expectChildToHaveDescription(0, 3);

keyDown("downArrow", 2);
testOutput += expectChildToHaveDescription(0, 1);

keyDown("\t");
testOutput += expectChildToHaveDescription(1, 2);

keyDown("1");
keyDown("0");
testOutput += expectChildToHaveDescription(1, 10);

...
```

2. To make the `-expected.txt` file more readable, it would be nice to log to the test output when you send an event. For example:

```
testOutput += "Simulating up arrow keypress.\n"
keyDown("upArrow");
// ...expectation...
// ...other keypresses / expectations...

testOutput += "Simulating tab key to move to next field in the date input.\n"
keyDown("\t");
...
```

Maybe you could have a helper function:

```
function keyDownAndLog(key) {
    testOutput += `Simulating ${key} keypress.\n`;
    keyDown(key);
}
```

3. Could you make this async? To do that, you'll need to:

Set `window.jsTestIsAsync = true;`, and call `finishJSTest();` after `debug(testOutput);`.

Update `expectChildToHaveDescription` to be like this:

```
async function expectChildToHaveDescription(...args...) {
    return expectAsync(...);
}
```

And because this function is async, you can only call it from async contexts, so I would move most of your test body into an async block like this:

```
setTimeout(async () => {
    ...
});
```

Any calls to `expectChildToHaveDescription` will have to be `await`'d. (Sorry if you know some / all of this).

```
testOutput += await expectChildToHaveDescription(...);
```

Making the test async is important because WebKit accessibility is multi-threaded and eventually consistent, so any dynamic changes won't be immediately reflected. You'll know if you've done it correctly if your test passes with and without the `--accessibility-isolated-tree` flag.

twilco commented on LayoutTests/accessibility/aria-label-valuetext-date-input.html:
```diff
@@ -0,0 +1,87 @@
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
+    var testOutput = "Tests that aria-label and aria-valuetext update correctly for date inputs.\n\n";
```
I don't think we actually test the aria-label yet 🤔 you should be able to get that with `platformValueForW3CName` (defined in `accessibility-helpers.js`).

What do you think about naming the test something like `date-input-aria-label-and-value.html`? Having date-input up-front reads a bit better IMO, but feel free to tweak that.

Really thorough set of testcases overall, nice job.

twilco commented on LayoutTests/accessibility/date-input-aria-label-and-value.html:
```diff
@@ -0,0 +1,100 @@
+            
+            async function expectChildToHaveARIALabel(childIndex, labelText) {
+                return expectAsync(`platformValueForW3CName(axDateFieldWrapper.childAtIndex(${childIndex}))`, labelText);
+            }
```
I don't think this function needs to be async (and it can use `expect` instead of `expectAsync`). The only things that need to be made async in WebKit accessibility layout tests are anything resulting from a dynamic page change via JS or DOM APIs.

Christop406 replied:
ok good to know for next time (about `expect` vs `expectAsync`)

twilco commented on LayoutTests/accessibility/date-input-aria-label-and-value.html:
```diff
@@ -0,0 +1,100 @@
+
+<input id="date-input" type="date" value="2022-02-02" />
+
+<script>
```
Everything inside this `<script>` can be moved back one indentation level.

twilco commented on LayoutTests/accessibility/date-input-aria-label-and-value.html:
```diff
@@ -0,0 +1,100 @@
+            testOutput += await expectChildToHaveARIAValueAttributes(2, 2013);
+
+            outputString("Manually setting the date input's value via JavaScript.")
+            dateInput.value = '2023-12-30';
```
This test mostly uses double quotes, but occasionally uses single quotes (like this spot). It would be nice to use double quotes consistently everywhere.

twilco reviewed (commented):
Almost there! Just a few style nits left.

I'm not sure why the style checker is failing either as I don't see you referencing the old test name anywhere. Does the style check fail locally? You should be able to use `Tools/Scripts/check-webkit-style`.

pxlcoder reviewed (approved):
Thank you for working on this!

If you are interested in another patch, I bet that input[type=time] could benefit from a similar treatment.

Commit:
AX: input[type=date] individual fields are announced as "group"
https://bugs.webkit.org/show_bug.cgi?id=251544
rdar://problem/104928713

Reviewed by Aditya Keerthi.

This fixes the bug by adding labels representing the "day", "month", and "year" inputs so that a screen reader has something to read,
instead of just "group" - it also switches the "role" attribute of the same divs to "spinbutton" to more accurately
reflect how they are used.

* LayoutTests/accessibility/date-input-aria-label-and-value-expected.txt: Added.
* LayoutTests/accessibility/date-input-aria-label-and-value.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/shadow/DateTimeFieldElements.cpp:
(WebCore::DateTimeDayFieldElement::create):
(WebCore::DateTimeMonthFieldElement::create):
(WebCore::DateTimeSymbolicMonthFieldElement::create):
(WebCore::DateTimeYearFieldElement::create):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
(WebCore::DateTimeNumericFieldElement::setEmptyValue):
(WebCore::DateTimeNumericFieldElement::setValueAsInteger):
(WebCore::DateTimeNumericFieldElement::setARIAValueAttributesWithInteger):
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.h:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXDateFieldMonthText):
(WebCore::AXDateFieldDayText):
(WebCore::AXDateFieldYearText):
* Source/WebCore/platform/LocalizedStrings.h:

Canonical link: https://commits.webkit.org/261123@main
